### ISSUE TITLE
Hibernate Validator: annotations on type parameters of nested classes are ignored (at least for value extractors)

### ISSUE THREAD

**Issue by yrodiere** [quarkusio/quarkus#20377]
### Describe the bug

When nested classes are involved, Hibernate Validator seems unable to detect annotations on type parameters of those classes, e.g. `NestedContainerType<@NotBlank String>` or `implements ValueExtractor<NestedContainerType<@ExtractedValue ?>>`

Moving the classes to the top-level fixes the problem.

### Expected behavior

When I declare a container type as a nested type, e.g. like this:

```
public class NestedContainerTypeCustomValueExtractorTest {

    // ...

    public static class NestedContainerType<T> {

        public T value;

    }

    @Singleton
    public static class NestedContainerClassValueExtractor
            implements ValueExtractor<NestedContainerType<@ExtractedValue ?>> {

        @Override
        public void extractValues(NestedContainerType<?> originalValue, ValueReceiver receiver) {
            receiver.value("someName", originalValue.value);
        }
    }

}
```

... then I expect Hibernate Validator to detect the `@ExtractedValue` annotation on the custom value extractor.

### Actual behavior

Exception on startup, suggesting that Hibernate Validator does not see the `@ExtractedValue` annotation:

```
java.lang.ExceptionInInitializerError
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:398)
        at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:215)
        at io.quarkus.test.QuarkusUnitTest.beforeAll(QuarkusUnitTest.java:509)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$10(ClassBasedTestDescriptor.java:381)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllCallbacks(ClassBasedTestDescriptor.java:381)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:205)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:80)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:148)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
        at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
        at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
        at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:188)
        at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:154)
        at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:128)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
        at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
Caused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:236)
        ... 46 more
Caused by: javax.validation.valueextraction.ValueExtractorDefinitionException: HV000203: Value extractor type io.quarkus.hibernate.validator.test.valueextractor.NestedContainerTypeCustomValueExtractorTest$NestedContainerClassValueExtractor fails to declare the extracted type parameter using @ExtractedValue.
        at org.hibernate.validator.internal.engine.valueextraction.ValueExtractorDescriptor.getExtractedTypeParameter(ValueExtractorDescriptor.java:108)
        at org.hibernate.validator.internal.engine.valueextraction.ValueExtractorDescriptor.<init>(ValueExtractorDescriptor.java:49)
        at org.hibernate.validator.internal.engine.AbstractConfigurationImpl.addValueExtractor(AbstractConfigurationImpl.java:271)
        at org.hibernate.validator.internal.engine.AbstractConfigurationImpl.addValueExtractor(AbstractConfigurationImpl.java:80)
        at io.quarkus.hibernate.validator.runtime.HibernateValidatorRecorder$2.created(HibernateValidatorRecorder.java:143)
        at io.quarkus.arc.runtime.ArcRecorder.initBeanContainer(ArcRecorder.java:70)
        at io.quarkus.deployment.steps.ArcProcessor$generateResources-1025303321.deploy_0(ArcProcessor$generateResources-1025303321.zig:128)
        at io.quarkus.deployment.steps.ArcProcessor$generateResources-1025303321.deploy(ArcProcessor$generateResources-1025303321.zig:40)
        at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:194)
        ... 46 more
```

Not this is not limited to `@ExtractedValue` or value extractors; if you don't declare a custom value extractor for the custom container type, constraint annotations on the extracted value (e.g. `NestedContainerType<@NotBlank String>`) will be ignored as well, instead of triggering an exception as expected because the value extractor cannot be found.

### How to Reproduce?

I'm going to add a disabled test as part of the PR for #20347. Look for `NestedContainerTypeCustomValueExtractorTest` in the codebase.

### Output of `java -version`

```
openjdk version "11.0.12" 2021-07-20
OpenJDK Runtime Environment Temurin-11.0.12+7 (build 11.0.12+7)
OpenJDK 64-Bit Server VM Temurin-11.0.12+7 (build 11.0.12+7, mixed mode)
```

### GraalVM version (if different from Java)

N/A (using JVM mode)

### Quarkus version or git rev

2.3.0.CR1, f0aba85

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

```
Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d)
Maven home: /home/yrodiere/tools/sdkman/candidates/maven/current
Java version: 11.0.12, vendor: Eclipse Foundation, runtime: /home/yrodiere/tools/java/jdk-11.0.12+7
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.13.6-200.fc34.x86_64", arch: "amd64", family: "unix"
```

**Comment by gsmet** [comment]
I debugged things a bit and it looks related to https://stackoverflow.com/questions/39952812/why-annotation-on-generic-type-argument-is-not-visible-for-nested-type .

We end up with the annotation in `AnnotatedTypeBaseImpl#allOnSameTargetTypeAnnotations` and it's not included in the API. I don't exactly know if we could find a workaround.

In any case, the issue is in HV, not in Quarkus.

@yrodiere if you end up having a clever idea, I'm all ears.

**Comment by yrodiere** [comment]
I managed to create a simple reproducer without Quarkus or Hibernate Validator: https://github.com/yrodiere/jdk-playground/blob/type-use-annotation-nested-type/src/test/java/org/hibernate/jdk/playground/MyTest.java

Turns out it's a JDK bug.

But... only in JDK 8 and 11. The bug seems fixed in JDK 17 and 18; at least my reproducer doesn't fail anymore with JDK 17 and 18.

I'll close this. To anyone with the same problem: try to upgrade to a newer JDK.
